core/prob_dist.py

```diff
@@ -100,3 +100,93 @@ def test_prob_creation_and_validation(self):
 
         # check if this works
         _ = ProbabilityDist(dist)
+
+
+def get_mean_log_prob(prob_dist: ProbabilityDist, data_block) -> float:
+    """computes the average log_probability of the input data_block given the probability distribution
+    prob_dist. This acts as a fundamental limit on what any compressor designed for distribution
+    prob_dist can achieve for the input data_block
+
+    Args:
+        prob_dist (ProbabilityDist): specified probability distribution used to compute log_probability
+        data_block (DataBlock): input for which avg log probability needs to be computed
+    """
+
+    # get avg log probability for the input
+    log_prob = 0
+    for s in data_block.data_list:
+        log_prob += prob_dist.log_probability(s)
+    avg_log_prob = log_prob / data_block.size
+    return avg_log_prob
+
+
+class Frequencies:
+    """
+    Wrapper around a frequency dict
+    NOTE: Frequencies is a typical way to represent probability distributions using integers
+    """
+
+    def __init__(self, freq_dict=None):
+
+        # NOTE: We use the fact that since python 3.6, dictionaries in python are
```

Comment:
where do we use this fact? Is it for the lists?

Reply:
I guess also for cumulative

Reply:
yes exactly

Reply:
Why do you need it to be OrderedDict? We don't really care order in which symbols were added to the dict, correct?

Reply:
We do need ordering. for example when we return the `freq_list` or the `cumulative_freq_list`.. we assume that the first frequency in the list corresponds to  the first symbol (for example).

Reply:
I think the important part in ordering is the consistency. The particular ordering according to insertion order is probably not that important because we will almost always insert all symbols at the beginning and later only update frequencies

Reply:
Got it. Thanks.